### PR TITLE
Post-check CSV file in built image

### PR DESCRIPTION
**Pull request by tkdchen**
* CLOUDBLD-2485

Signed-off-by: Chenxiong Qi <cqi@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- n/a JSON/YAML configuration changes are updated in the relevant schema
- n/a Changes to metadata also update the documentation for the metadata
- n/a Pull request has a link to an osbs-docs PR for user documentation updates
- n/a New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/utils/operator.py [hunk 1]
```diff
@@ -314,6 +314,11 @@ def __init__(self, path, data, pullspec_heuristic=default_pullspec_heuristic):
         self.data = data
         self._pullspec_heuristic = pullspec_heuristic
 
+    @property
+    def md5sum(self):
```
Comment (ben-alkov) [inline atomic_reactor/utils/operator.py:318]: If any of the concern about verifying that the CSVs match involves potential security implications from e.g. spoofing CSVs, you will definitely want to use something other than MD5 - it is no longer considered to be a "crypto" hash (it's not really any better than e.g. XOR).

Comment (tkdchen) [reply atomic_reactor/utils/operator.py]: @ben-alkov Thanks for this info. Fixed.

File: atomic_reactor/util.py [hunk 2]
```diff
@@ -1819,3 +1819,12 @@ def _enclose_pullable(self):
 
             if self._should_enclose(image):
                 image.enclose(self._organization)
+
+
+def sha256sum(s, abbrev_len=0, prefix=False):
+    prefix_ = 'sha256:' if prefix else ''
+    data = s.encode() if type(s) == str else s
```
Comment (ben-alkov) [inline atomic_reactor/util.py:1826]: @tkdchen; To satisfy my own curiosity - aren't Py 3 strings unicode by default?

File: atomic_reactor/plugins/post_export_operator_manifests.py [hunk 3]
```diff
@@ -133,3 +135,27 @@ def run(self):
...
         return manifests_zipfile_path
+
+    def _verify_csv(self, manifests_path):
+        """Verify the CSV file from the built image
+
+        :param str manifests_path:
+        :raises: ValueError if more than one CSV files are found, or the single
+            CSV is different from the one in the repo (compared by hash digest).
+        """
+        image_csv = OperatorManifest.from_directory(manifests_path).csv
```
Comment (rcerven) [inline atomic_reactor/plugins/post_export_operator_manifests.py:146]: this will raise https://github.com/containerbuildsystem/atomic-reactor/blob/master/atomic_reactor/utils/operator.py#L597
Operator bundle may contain only 1 CSV file, but contains more

if in the OUTPUT directory is more CSV files

so maybe catch it and reraise with : in the output operator bundle, are more CSV files ....

File: atomic_reactor/util.py (hunk 2 above)
Comment (tkdchen) [reply atomic_reactor/util.py:1826]: @ben-alkov This is refactored from `make_digest`, which calls `to_bytes` to accept either a str or bytes string and ensure a bytes string is passed to `hashlib.sha256`.

File: atomic_reactor/plugins/post_export_operator_manifests.py [hunk 4]
```diff
@@ -125,11 +127,41 @@ def run(self):
...
+        """
+        try:
+            image_csv = OperatorManifest.from_directory(manifests_path).csv
+        except ValueError as e:
+            if 'Missing ClusterServiceVersion' in str(e):
+                raise ValueError('No ClusterServiceVersion in output operator bundle.')
+            if 'Operator bundle may contain only 1 CSV file' in str(e):
+                raise ValueError(
+                    'The output operator bundle contains more than one CSV files, '
+                    'but only 1 should be included.')
```
Comment (MartinBasti) [inline atomic_reactor/plugins/post_export_operator_manifests.py:151]: Can we have something like this instead searching in error messages?

```suggestion
            raise ValueError("Operator manifests check in built image failed: {e}")
```

File: atomic_reactor/util.py (hunk 2 above)
Comment (MartinBasti) [inline atomic_reactor/util.py:1826]:
```suggestion
    data = s.encode() if isinstance(s, str) else s
```

File: atomic_reactor/plugins/post_export_operator_manifests.py [hunk 5]
```diff
@@ -125,11 +127,41 @@ def run(self):
...
+            raise
+
+        manifests_dir = os.path.join(
```
Comment (MartinBasti) [inline atomic_reactor/plugins/post_export_operator_manifests.py:154]: In `pre_pin_operator_digest.PinOperatorDigestPlugin_get_operator_manifest` we have code for making safe path to load data (and fail otherwise). Would you mind to factor it out and use in both plugins for consistency?

Basically this code:
```python
        repo_dir = os.path.realpath(self.workflow.source.path)
        manifests_rel_path = self.user_config["manifests_dir"]

        manifests_dir = os.path.realpath(os.path.join(repo_dir, manifests_rel_path))
        if not manifests_dir.startswith(repo_dir):
            raise RuntimeError("manifests_dir points outside of cloned repository")
```

Comment (tkdchen) [reply atomic_reactor/plugins/post_export_operator_manifests.py]: Created a new property `Source.manifests_dir`

**Comment by ben-alkov** [comment]
> This pull request **introduces 1 alert** when merging [91a4943](https://github.com/containerbuildsystem/atomic-reactor/commit/91a49432a41a0f73b0b3ec78c066d0eba5e41e2a) into [2ab5516](https://github.com/containerbuildsystem/atomic-reactor/commit/2ab5516cf569015ad794c709ea9869a40a61dabd) - [view on LGTM.com](https://lgtm.com/projects/g/containerbuildsystem/atomic-reactor/rev/pr-97e0ca75f5dc950f58a09c6ba4b8fcab420a2230)
> 
> **new alerts:**
> 
>     * 1 for Unused import

@tkdchen; Specifically "atomic_reactor/plugins/pre_pin_operator_digest.py:10: [W0611(unused-import), ] Unused import os.path)"

**Comment by ben-alkov** [comment]
@tkdchen; Also, unit tests seem to be broken.

**Comment by MartinBasti** [comment]
Unittests are still failing

**Comment by rcerven** [comment]
code looks good, just unit tests now

File: atomic_reactor/plugins/post_export_operator_manifests.py [hunk 6]
```diff
@@ -117,6 +117,8 @@ def run(self):
         tar_archive.extractall(manifests_archive_dir)
         manifests_path = os.path.join(manifests_archive_dir, MANIFESTS_DIR_NAME)
 
+        self._verify_csv(manifests_path)
```
Comment (MartinBasti) [inline atomic_reactor/plugins/post_export_operator_manifests.py:120]: This verification should run only for operator bundles not for appregistry manifests

Comment (tkdchen) [reply atomic_reactor/plugins/post_export_operator_manifests.py:120]: @MartinBasti Any hint how to determine current built image is an operator bundle not appregistry manifests? Thanks.

Comment (MartinBasti) [reply atomic_reactor/plugins/post_export_operator_manifests.py:120]: https://github.com/containerbuildsystem/atomic-reactor/blob/2ab5516cf569015ad794c709ea9869a40a61dabd/atomic_reactor/util.py#L1542

File: atomic_reactor/plugins/post_export_operator_manifests.py [hunk 7]
```diff
@@ -125,11 +128,31 @@ def run(self):
                     filepath = os.path.join(filedir, f)
                     archive.write(os.path.join(root, f), filepath, zipfile.ZIP_DEFLATED)
             manifest_files = archive.namelist()
-            if not manifest_files:
```
Comment (MartinBasti) [inline atomic_reactor/plugins/post_export_operator_manifests.py:128]: Why was this removed?

Comment (tkdchen) [reply atomic_reactor/plugins/post_export_operator_manifests.py]: Added it back. It was removed due to the `_verify_csv` was called without checking the operator bundle manifests. Thanks for pointing out this.

**Review by MartinBasti — APPROVED** [review]
:rocket:  nice, thanks

**Comment by MartinBasti** [comment]
bump for review

**Review by rcerven — APPROVED** [review]
